File: 1-1000/0956-tallest-billboard/0956-tallest-billboard.cpp
class Solution {
public:    
    int tallestBillboard(vector<int>& rods) {
        int total = 0;
        static int dp[5001], temp[5001];
        for(int i = 1; i < 5001; ++i) dp[i] = INT_MIN;
        dp[0] = 0;
        for(int j = 0, n = rods.size(); j < n; ++j){
            int r = rods[j];
            total += r;
            for(int i = 0; i <= total; ++i) temp[i] = dp[i];
            for(int diff = 0; diff + r <= total; ++diff){
                int dp1 = diff + r, dp2 = abs(diff - r);
                dp[dp1] = max(dp[dp1] , temp[diff]);
                dp[dp2] = max(dp[dp2] , temp[diff] + min(diff , r));
            }
        }
        return dp[0];
    }
};

File: 1-1000/0956-tallest-billboard/0956-tallest-billboard.java
class Solution {
    public int tallestBillboard(int[] arr) {
        int sum = 0;
        for (int i : arr) {
            sum += i;
        }
        int[] cur = new int[sum + 1];
        Arrays.fill(cur, -1);
        cur[0] = 0;
        for (int i : arr) {
            int[] dp = cur.clone();
            for (int diff = 0; diff < dp.length; diff++) {
                if (cur[diff] == -1) {
                    continue;
                }
                // if (diff + i <= sum) {
                    dp[diff + i] = Math.max(dp[diff + i], cur[diff] + i);
                // } 
                int newDiff = Math.abs(diff - i);
                int newHeight = Math.max(cur[diff], cur[diff] - diff + i);
                dp[newDiff] = Math.max(dp[newDiff], newHeight);
            }
            cur = dp;
        }
        return cur[0] == -1 ? 0 : cur[0];
    }
}